### PR TITLE
Factory closure and callable will work identical

### LINKED ISSUE #61: Factory closure to work as callable
Right now, using callable for factory is like this (I removed options for brevity):

```php
public function buildForm(FormBuilderInterface $builder, array $options): void
{
    $builder->add('campaign', EntityType::class);       
    $builder->add('invoicedToClient', DateType::class);
}

public function configureOptions(OptionsResolver $resolver): void
{
    $resolver->setDefaults([
        'factory' => [$this, 'factory'],
    ]);
}

public function factory(Campaign $campaign, DateTime $invoicedToClient): Invoice
{
    return new Invoice($campaign, $invoicedToClient);
}
```

The signature of ``factory()`` method must match names of form fields which helps in finding errors. 

Typehinting parameters will also help static analysis tools; example is if user adds new dependency to Invoice::__construct() method, these tools will find issue in this ``InvoiceType`` form. 

Example:

```php
class Invoice
{
	public function __construct(Campaign $x, DateTime $y, bool $c)
	{...}
}
```

Unlike callables, [closures](https://github.com/sensiolabs-de/rich-model-forms-bundle/blob/master/docs/factory_value_object.md#initializing-objects)  will receive form data as array.

So creating instance of Invoice would prevent static analysis:

```php
'factory' => function (array $data) {
    return new Invoice($data['campaign', $data['invoicedToClient'])
}
```

---

My suggestion is to change this decision and allow Closures to work the same as callables. If accepted, above example would be:

```php
'factory' => function (Campaign $campaign, DateTime $invoicedToClient) {
	return new Invoice($campaign, $invoicedToClient);
},
```

Another big advantage is that this way, we could use form normalizers and construct data object in different way, based on some other option.

So if param ``$c`` is supposed to be injected from controller, normalizer would be this:

```php
public function configureOptions(OptionsResolver $resolver): void
{
    $resolver->setRequired('some_param');
 	$resolver->setNormalizer('factory', function (Options $options) {
		$someParam = $options['some_param'];

		return function ($campaign, $invoicedToClient) use ($someParam) {
			return new Invoice($campaign, $invoicedToClient, $someParam);
		}

		// or with upcoming arrow functions:
		// return fn($campaign, $invoicedToClient) => new Invoice($campaign, $invoicedToClient, $someParam);
	})
}

```

With callables, we cannot use normalizers because other options cannot be passed to it.

### PR DESCRIPTION
**Pull request by vladyslavstartsev**
made behavior of factory closures and callables identical
close #61

any feedback is welcome

### CODE REVIEW

**Comment by xabbuh** [comment]
Thank you for working on this @vladyslavstartsev. The implementation looks quite good to me. Can you please have a look at the build failures (seems to be CS and PhpStan related)? And could you also add an entry to the changelog to describe the BC break that we will introduce with this change?

**Comment by vladyslavstartsev** [comment]
@xabbuh Done.

By the way, it's related to other PR I think, but can you release 0.7.0 only after all phpstan errors (baseline) are resolved?

**Comment by vladyslavstartsev** [comment]
also I want to here feedback from @zmitic

**Comment by zmitic** [comment]
@vladyslavstartsev Thank you, I will take another look this week. Can't do it now as my code depends on different bundle (quick&dirty patch with no tests).

**Comment by xabbuh** [comment]
@vladyslavstartsev Can you rebase here? I have made some changes to get rid of the PhpStan baseline entries.

**Comment by vladyslavstartsev** [comment]
@xabbuh done!

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,8 @@ CHANGELOG
 -----
 
 * Allow the bundle to be used on PHP 8.
+* [BC BREAK] the behaviour of `factory` option when `\Closure` provided changed.
```
Comment (xabbuh) [inline CHANGELOG.md:8]: I think I would reword the entry a bit like this:

```markdown
* [BC BREAK] When closures (anonymous functions) are used for the `factory` option, the submitted
  form data is no longer passed as an array, but each child form is passed as a single argument
  instead (use the `factory_argument` option of the child forms if form names and argument names
  of the closure differ).
```

File: src/Instantiator/ObjectInstantiator.php [hunk 2]
```diff
@@ -36,30 +36,28 @@ public function __construct($factory)
      */
     public function instantiateObject()
     {
-        if ($this->factory instanceof \Closure) {
-            return ($this->factory)($this->getData());
-        }
-
-        if (\is_string($this->factory)) {
+        if (\is_string($this->factory) && class_exists($this->factory)) {
```
Comment (xabbuh) [inline src/Instantiator/ObjectInstantiator.php:39]: Do we really need this additional check?

File: src/Instantiator/ObjectInstantiator.php [hunk 3]
```diff
@@ -36,30 +36,28 @@ public function __construct($factory)
...
             $factoryMethod = (new \ReflectionClass($this->factory))->getConstructor();
-
+            $factoryMethodAsString = $this->factory.'::__construct';
```
Comment (xabbuh) [inline src/Instantiator/ObjectInstantiator.php:41]: Can you please revert this change to keep the diff as small as possible?

**Comment by xabbuh** [comment]
@vladyslavstartsev The changes look very nice from my POV. 👍 I have just left some minor nitpicking comments.

File: src/Instantiator/ObjectInstantiator.php (hunk 2 above)
Comment (vladyslavstartsev) [reply src/Instantiator/ObjectInstantiator.php:39]: yes, since we don't know the type of `$this->factory` it can be `class-string|\Closure|(callable&array)` AND `class_exists` expects string.

It can be done the way I did OR the way below (move `class_exists` check to the end)

```php
        if (\is_array($this->factory) && \is_callable($this->factory)) {
            $class = \is_object($this->factory[0]) ? \get_class($this->factory[0]) : $this->factory[0];
            $factoryMethod = (new \ReflectionMethod($class, $this->factory[1]));
            $factoryMethodAsString = $class.'::'.$this->factory[1];
            if (!$factoryMethod->isPublic()) {
                throw new TransformationFailedException(sprintf('The factory method %s() is not public.', $factoryMethodAsString));
            }
        } elseif ($this->factory instanceof \Closure) {
            $factoryMethod = new \ReflectionFunction($this->factory);
        } elseif (class_exists($this->factory)) {
            $factoryMethod = (new \ReflectionClass($this->factory))->getConstructor();
            $factoryMethodAsString = $this->factory.'::__construct';
            if (null === $factoryMethod) {
                throw new TransformationFailedException(sprintf('The class "%s" used as a factory does not have a constructor.', $this->factory));
            }
            if (!$factoryMethod->isPublic()) {
                throw new TransformationFailedException(sprintf('The factory method %s() is not public.', $factoryMethodAsString));
            }
        } else {
            return $this->getData();
        }
```

BUT since you're saying 
> keep the diff as small as possible
I will leave it the way I did

**Comment by vladyslavstartsev** [comment]
thanks @xabbuh !

Waiting for you @zmitic (not for a lot time, I want to see this feature in my code)

File: src/Instantiator/ObjectInstantiator.php (hunk 2 above)
Comment (xabbuh) [reply src/Instantiator/ObjectInstantiator.php:43]: I was mainly wondering if we were missing that before (no need to revert in that case by the way I am just trying to understand :)) or if it was needed because of the other changes here.

File: src/Instantiator/ObjectInstantiator.php [hunk 4]
```diff
@@ -14,6 +14,10 @@
 
 namespace SensioLabs\RichModelForms\Instantiator;
 
+use Closure;
```
Comment (xabbuh) [inline src/Instantiator/ObjectInstantiator.php:17]: Can you revert the `use` statements for classes from the global namespace please? :)

**Comment by xabbuh** [comment]
If nothing serious pops up, I will probably merge this PR and tag a new release on Friday.

File: src/Instantiator/ObjectInstantiator.php (hunk 2 above)
Comment (vladyslavstartsev) [reply src/Instantiator/ObjectInstantiator.php:43]: PHPStan is yelling at me if I do

```
if (class_exists($this->factory)) {
```

File: src/Instantiator/ObjectInstantiator.php (hunk 4 above)
Comment (vladyslavstartsev) [reply src/Instantiator/ObjectInstantiator.php:17]: @xabbuh you should probably force this at `php-cs-fixer` level.

File: src/Instantiator/ObjectInstantiator.php (hunk 2 above)
Comment (xabbuh) [reply src/Instantiator/ObjectInstantiator.php:39]: Well, but why do we need the `class_exists()` call at all? Does anything fail without it?

Comment (vladyslavstartsev) [reply src/Instantiator/ObjectInstantiator.php]: Nothing fails.

Thanks for suggestion

**Comment by xabbuh** [comment]
Thank you @vladyslavstartsev.

**Comment by xabbuh** [comment]
I have released 0.7 with this PR. Thank you @vladyslavstartsev.
